### DISCUSSION TITLE
Unable to access mcp23017

### DISCUSSION THREAD

**Discussion by fraygon** [jpcornil-git/HA-mcp23017#8] [General]
Looking for the replacement of the I2C integration in HA that will be lost I was advised to adopt yours. Unfortunately it did not go well.
I have implemented the I2C ports as the integration of H.A. but has. gives me this error that I can't interpret:
"Error: Unable to access mcp23017-0x20 ([Errno 2] No such file or directory: '/ dev / i2c-1')"
Could you tell me what I did wrong?
thank you

**Comment by jpcornil-git** [comment]
What type of homeassistant installation are you using ? (HA OS/Container/Core, cfr. https://www.home-assistant.io/installation/)
What is your platform ? (RPi ?)
How did you enabled i2c ? ("what do you mean with I2C ports as the integration" ?)

**Comment by fraygon** [comment]
I use Hassio with RPi 4 and H.A. runs on an SD.
I have enabled the I2C ports as suggested in HA's "Common Tasks - Operating System". I had ports enabled before and everything worked fine with H.A.'s "MCP23017 I / O Expander" integration.

**Comment by jpcornil-git** [comment]
The error indicates that /dev/i2c-1 (i2c bus) is not available.
Given you are using HAOS, do you have the ssh addon installed, if not install it, run following commands and report results

Check presence of /dev/i2c-1 
```
ls -l /dev/ | grep i2c
```
If present then install i2detect and check devices connected to /dev/i2c-1
```
apk add i2c-tools
i2cdetect -y 1
```

**Comment by jpcornil-git** [comment]
For information, I configured a spare RPIv4 with fresh HAOS 7.5 running core 2022.3.8:
- enabled i2c (manually uncommented  dtparam=i2c_arm=on in config.txt and added i2c-dev to modules/rpi-i2c.conf but other methods are possible https://www.home-assistant.io/common-tasks/os#enable-i2c)
- installed HACS
- installed mcp23017 from HACS

... and everything is working fine

**Reply by fraygon** [reply]
Thank you for your help. I always ran those commands and it always worked.
The difference (I understand) with respect to your sequence is that HACS is installed but also MCP23017 was already installed when I tried to enable the I2C ports (for the record are the usual GPIO2 SDA and GPIO3 SCL?). Could it be useful to uninstall MCP23017, repeat the procedure to enable I2C and reinstall MCP23017?
Sorry for the length of time. Thank you

**Comment by jpcornil-git** [comment]
I2c-1 is mapped on GPIO 2 and 3 (pin 3 and 5 of the RPi  [40-pin header](https://www.raspberrypi.com/documentation/computers/os.html#gpio-and-the-40-pin-header).

Presence of HACS/mcp23017 has not impact on enabling i2c (or the other way around).

Could you run from HA ssh/console
```
ls -l /dev/ | grep i2c
```
and report result ? (this should confirm that i2c is nor enabled)

**Comment by fraygon** [comment]
![Immagine 2022-04-01 185436](https://user-images.githubusercontent.com/69857110/161308300-73c4a5af-2290-406a-a974-c01f41a94283.png)
i don't know if it's correct as i wrote the command but i don't interpret the result.
I'm sorry to waste your time.

**Comment by jpcornil-git** [comment]
If i2c was effectively enabled you should have seen:
```bash
[core-ssh ~]$ ls -l /dev/ | grep i2c
crw-------    1 root     root       89,   1 Apr  1 12:14 i2c-1
```
-> something went wrong/was incorrect when you enabled i2c

Could you describe what you did as exactly as possible to set it up ?

**Comment by fraygon** [comment]
I extracted the SD. Through a pc win open the files. in Hassos boot I created the CONFIG folder. And then all the sequence indicated in
https://www.home-assistant.io/common-tasks/os#enable-i2c
I will try to redo the operation, maybe it will be better. I'm not familiar with SSH but to check if i2c is enabled okay how did I do?

**Comment by jpcornil-git** [comment]
Just run the same "ls -l /dev/ | grep i2c" command, it should return what I reported for mine in https://github.com/jpcornil-git/HA-mcp23017/discussions/8#discussioncomment-2487360 (while yours return nothing/is empty)

**Comment by jpcornil-git** [comment]
For information, I did a fresh HAOS install, followed https://www.home-assistant.io/common-tasks/os#enable-i2c using a windows machine to create directories & files and had no issues to setup i2c either.

Note that with windows you should be careful when creating a new file (rpi-i2c.conf here) as it may add a hidden .txt extension -> open a command prompt and check. If you see  rpi-i2c.conf.txt rename it rpi-i2c.conf.

Once enabled (first system boot to update the configuration followed by a 2nd one to pick it up), you should also see the i2c interface in the hardware list (Supervisor -> thre-dots -> hardware)

![image](https://user-images.githubusercontent.com/40644331/161375342-dccbe17e-28ca-4580-b469-338db8e8e155.png)

Once it is there, install HACS/mcp23017 and it should work as well

Note also that it is enough to uncomment or add the following line in config.txt ("dtparam=i2c_vc=on" enable VideoCore's i2c for the CSI/camera interface, not required here)
```
dtparam=i2c_arm=on
```

**Comment by fraygon** [comment]
With pleasure I inform you that everything works.
I basically did these things:
modified the rpi-i2c.conf.txt file by removing the "txt" suffix.
After a couple of reboots via "settings" "reboot" nothing has changed.
I then completely shut down everything and rebooted twice and eventually the MCP23017 entities appeared.
Thanks for the help.
Please don't abandon the integration too: my home automation is based on self-built pcbs.
Another disappointment and goodbye to Home Assistant.

**Comment by jpcornil-git** [comment]
Thanks for the update/great to hear this :-)

Ultimately the issue was à windows issue (hidden txt suffix). You have to reboot linux/the whole platform twice (pick up the configuration change and apply them), homeassistant is not enough/nothing to blame here as it csn only make use of what is offered by the underlying OS.

Enjoy now :-)
